### PR TITLE
Implementing W_pdist Test

### PR DESCRIPTION
**Pull request by jeremyleung521**
Adding test_w_pdist.py functional test and ref file (pdist_ref.h5).  Sorry about the previous push accident/mistake!

### CODE REVIEW

File: tests/test_tools/test_w_pdist.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
+import os
+import shutil
+from .h5diff import H5Diff
+import unittest
+from westpa.cli.tools.w_pdist import entrypoint
```
Comment (synapticarbors) [inline tests/test_tools/test_w_pdist.py:5]:
```suggestion
from westpa.cli.tools.w_pdist import entry_point
```

File: tests/test_tools/test_w_pdist.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    test_name = 'W_PDIST'
+
+    def test_run_w_pdist(self):
+        '''Testing if w_pdist runs as expected and the pdist.h5 file looks good.'''
+
+        ref_dir = os.path.join(os.path.dirname(__file__), 'refs')
+        shutil.copy2(os.path.join(ref_dir, 'west.cfg'), './')
+        shutil.copy2(os.path.join(ref_dir, 'west.h5'), './')
+
+        entrypoint()
```
Comment (synapticarbors) [inline tests/test_tools/test_w_pdist.py:20]:
```suggestion
        entry_point()
```

**Comment by jeremyleung521** [comment]
I'll close this for now so I'm not bombarding the mailing list with new checks. Will reopen when it's ready.

**Comment by synapticarbors** [comment]
@jeremyleung521 -- just as a note, you can always run the test suite locally using `pytest -v tests` from the top of the repository. This is pretty much what the CI runs, so if you get it to pass locally, you're less likely to be surprised by what happens once you push it here.  You can also test a single file if you don't want to test the entire suite using:

```
pytest -v tests/test_tools/test_w_pdist.py
```

**Comment by jeremyleung521** [comment]
Okay. The test is done. 

One odd issue I had:
If I put "last_iter=None" (the default) and let the code try to figure out the last iteration number, it fails while looking for wemd_current_iteration, a legacy(?) west.h5 attribute. It might be related to _rc.py not being called, causing some issues with the data manager (iter_range.iterRangeSelection or data_manager.current_iteration). This error is not present if I run w_pdist from the command line. Since it's not directly related to w_pdist and outside the scope of this test, I circumvented it by directly indicating "last_iter=50".

> @jeremyleung521 -- just as a note, you can always run the test suite locally using `pytest -v tests` from the top of the repository. This is pretty much what the CI runs, so if you get it to pass locally, you're less likely to be surprised by what happens once you push it here. You can also test a single file if you don't want to test the entire suite using:
> 
> ```
> pytest -v tests/test_tools/test_w_pdist.py
> ```

Thank you, Josh! That definitely helped a lot during the process of rewriting the test!

**Comment by synapticarbors** [comment]
@jeremyleung521 -- I believe the `wemd_current_iteration` is a legacy attribute, and should only attempted to be accessed if `west_current_iteration` is not there:

https://github.com/westpa/westpa/blob/2f418129d50e34def9947e0a8100278c5b2c1c6d/src/westpa/core/data_manager.py#L317-L325

You didn't post the traceback, but what I'm guessing is happening is that `west_current_iteration` is missing from the h5 file and then it tries to fall back to checking `wemd_current_iteration`, but that isn't there either so it fails. Looking at the `west.h5` in `refs, it is missing that attribute:

```python
In [5]: f = h5py.File('/Users/xyz/Downloads/west.h5', 'r')

In [6]: f.keys()
Out[6]: <KeysViewHDF5 ['bin_topologies', 'ibstates', 'iterations', 'summary', 'tstates']>
```

@atbogetti -- How was the `west.h5` file in refs created?

From briefly reading through the code, it looks like the `current_iteration` setter, that actually changes the `west_current_iteration` attribute in the h5 only ever gets set in the following places:

```bash
$ rg "current_iteration ="
src/westpa/core/sim_manager.py
338:        data_manager.current_iteration = 1

src/westpa/core/data_manager.py
366:            self.current_iteration = 0

src/westpa/cli/core/w_fork.py
130:    dm_new.current_iteration = 1

src/westpa/cli/core/w_truncate.py
38:    dm.current_iteration = n_iter - 1
```

I don't actually see the value being updated to any new values, so I'm not sure what this functionality is actually meant for, unless I'm missing something. 

I would like to know what is different when `w_pdist` is called from the command line in terms of how it is setting the value. 

@mczwier -- Any thoughts on this?

**Comment by jeremyleung521** [comment]
Traceback: Unfortunately h5py hides a lot of the output information.

```
../../src/westpa/cli/tools/w_pdist.py:487: in entry_point
    WPDist().main()
../../src/westpa/tools/core.py:165: in main
    self.make_parser_and_process()
../../src/westpa/tools/core.py:143: in make_parser_and_process
    self.process_all_args(args)
../../src/westpa/tools/core.py:63: in process_all_args
    fn(self, args)
../../src/westpa/cli/tools/w_pdist.py:282: in process_args
    self.iter_range.process_args(args)
../../src/westpa/tools/iter_range.py:86: in process_args
    self.iter_stop = (self.data_manager or westpa.rc.get_data_manager()).current_iteration
../../src/westpa/tools/data_reader.py:61: in __getattr__
    return getattr(self.data_manager, key)
../../src/westpa/core/data_manager.py:325: in current_iteration
    return int(self.we_h5file['/'].attrs['wemd_current_iteration'])
h5py/_objects.pyx:54: in h5py._objects.with_phil.wrapper
    ???
h5py/_objects.pyx:55: in h5py._objects.with_phil.wrapper
    ???
../../../../anaconda3/envs/westpa2/lib/python3.7/site-packages/h5py/_hl/attrs.py:60: in __getitem__
    attr = h5a.open(self._id, self._e(name))
h5py/_objects.pyx:54: in h5py._objects.with_phil.wrapper
    ???
h5py/_objects.pyx:55: in h5py._objects.with_phil.wrapper
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   KeyError: "Can't open attribute (can't locate attribute: 'wemd_current_iteration')"

h5py/h5a.pyx:77: KeyError

```

west_current_iteration can be found using the following and it is indeed 51, eliminating the possibility of a corrupted west.h5:
```python
In [2]: import h5py                                                             

In [3]: f = h5py.File('west.h5', 'r') 

In [4]: f.attrs.keys()                                                          
Out[4]: <KeysViewHDF5 ['west_current_iteration', 'west_file_format_version', 'west_iter_prec', 'westpa_fileformat_version', 'westpa_iter_prec']>

In [5]: f.attrs['west_current_iteration']                                       
Out[5]: 51
```

Here's the input for w_pdist from the command line, as much as I can extract.
[w_pdist_debug.log](https://github.com/westpa/westpa/files/5143107/w_pdist_debug.log)

**Comment by synapticarbors** [comment]
Ok, I can see that I was looking at `f.keys()` instead of the correct `f.attrs.keys()` and that the refs file contains `west_current_iteration`. That said, from your traceback, it looks like it is not seeing it properly since it hits the logical branch looking for `wemd_current_iteration`:

```
../../src/westpa/core/data_manager.py:325: in current_iteration
    return int(self.we_h5file['/'].attrs['wemd_current_iteration'])
```

It should only look there if `west_current_iteration` isn't in the keys for the h5 file's attributes. So something is going on that still doesn't look right to me.

**Comment by synapticarbors** [comment]
Also, I realized that the `current_iteration` is being updated in the sim_manager, so that's fine:

https://github.com/westpa/westpa/blob/2f418129d50e34def9947e0a8100278c5b2c1c6d/src/westpa/core/sim_manager.py#L682

my grepping just was looking for the wrong thing.

**Comment by atbogetti** [comment]
To fix that wemd_current_iteration issue you can remove __init__.py from the test_tools directory.  That worked for my tests.  Thanks @jdrusso for the suggestion!

**Comment by jeremyleung521** [comment]
> To fix that wemd_current_iteration issue you can remove `__init__.py` from the test_tools directory. That worked for my tests. Thanks @jdrusso for the suggestion!

Just for the records, deleting `__init__.py` will also break some of the existing tests so make sure to remove relative importing for all tests in the test_tools directory.

e.g. change `from .h5diff import H5Diff` to `from h5diff import H5Diff`

Thanks @atbogetti for the fix.

**Comment by jeremyleung521** [comment]
Added fixtures to test_w_pdist.py + some light organization of refs folder. Had to rename some file names in existing tests to prevent it from interfering with how the fixtures work.

**Comment by jeremyleung521** [comment]
Merged into #136.

**Pull request closed without merging**